docs/proposals/THV-2301-refactor-mcpregistry-controller.md

```diff
@@ -0,0 +1,315 @@
+
+### Design Constraints
+
+This refactoring maintains feature parity without introducing new capabilities:
+
+- **API stability**: The `MCPRegistry` CRD spec remains unchanged
+- **Status evolution**: Status field changes will be documented if required
+- **CLI extension**: New runtime flags may be added to `thv-registry-api`
+- **Storage**: File-based storage will be implemented; database storage is out of scope
+
+## Current status
+### Data source configuration
+`MCPRegistry` resources define data sources and filtering rules in the `spec.source` and `spec.filter` fields:
+```yaml
+spec:
+  source:
+    type: configmap
+    configmap:
+      name: minimal-registry-data
+  syncPolicy:
+    interval: "30m"
+  filter:
+    tags:
+      include: ["database", "production"]
+      exclude: ["experimental", "deprecated", "beta"]
+```
+### Current Implementation
+The `MCPRegistry` controller implements data source management in these operator packages:
+| Package| Purpose |
+|--------|---------|
+|`cmd/thv-operator/pkg/sources`|`SourceHandler` interface and implementations for `ConfigMap`, Git, and Registry API sources; StorageManager implementation using `ConfigMap` |
+|`cmd/thv-operator/pkg/httpclient`|HTTP client for Registry API sources|
+|`cmd/thv-operator/pkg/git`|Git repository operations|
+|`cmd/thv-operator/pkg/sync`|`Manager` interface for periodic and on-demand synchronization|
+|`cmd/thv-operator/pkg/filtering`|`FilterService` interface and implementation for tag and name filtering|
+
+These packages run exclusively within the operator controller with no external dependencies.
+
+### Controller Responsibilities
+The controller currently handles two distinct concerns:
+1. Deployment management: Creates and maintains Registry API server Deployments and Services
+1. Data synchronization: Fetches, filters, and stores registry data in `ConfigMap`s
+
+After deployment, the controller does not communicate with the Registry API server—it only manages the `ConfigMap` containing registry data.
+
+### Status Reporting
+The MCPRegistry status tracks deployment, synchronization and storage state information.
+
+#### API deployment status
+```yaml
+apiStatus:
+  endpoint: http://thv-git-api.toolhive-system:8080
+  phase: Ready
+  readySince: "2025-10-22T08:30:38Z"
+```
+#### Synchronization status  
+```yaml
+syncStatus:
+  phase: Complete
+  lastSyncTime: "2025-10-22T08:30:27Z"
+  serverCount: 88
+  lastSyncHash: c013b6b36286ab...
+```
+#### Storage reference
+```yaml
+storageRef:
+  type: configmap
+  configMapRef:
+    name: thv-git-registry-storage
+```
+#### Standard conditions
+```yaml
+conditions:
+  - type: APIReady
+    status: "True"
+    lastTransitionTime: "2025-10-22T08:30:38Z"
+```
+
+## Refactoring Proposal
+
+### Component Responsibilities
+
+**Controller:**
+- Deploy and manage Registry API (`Deployment`, `Service`)
+- Create and update configuration `ConfigMap` from `MCPRegistry` spec
```

Comment:
this is the configuration of the registry server right?

Reply:
Yes, the config of the server, populated by the controller and mounted to the server deployment